File: pe-core/src/main/java/org/saar/core/model/Meshes.java
package org.saar.core.model;

import org.saar.core.model.vertex.ModelBufferWriter;
import org.saar.core.model.vertex.ModelIndices;
import org.saar.lwjgl.opengl.constants.DataType;
import org.saar.lwjgl.opengl.constants.RenderMode;

import java.util.List;

public final class Meshes {

    private Meshes() {
        throw new AssertionError("Cannot create instance of class "
                + getClass().getSimpleName());
    }

    public static <T extends Vertex> Mesh arraysModel(
            RenderMode renderMode, ModelBufferWriter<T> writer, List<T> vertices) {
        writer.write(vertices);
        return new ArraysMesh(writer.toVao(), renderMode, vertices.size());
    }

    public static <T extends Vertex> Mesh elementsModel(
            RenderMode renderMode, ModelBufferWriter<T> writer,
            ModelIndices indices, List<T> vertices) {
        writer.write(vertices);
        writer.write(indices);
        return new ElementsMesh(writer.toVao(), renderMode, indices.count(), DataType.U_INT);
    }
}

File: pe-core/src/main/java/org/saar/core/mesh/Meshes.kt
package org.saar.core.mesh

import org.saar.core.mesh.prototype.IndexedVertexMeshPrototype
import org.saar.core.mesh.prototype.InstancedIndexedVertexMeshPrototype
import org.saar.core.mesh.prototype.InstancedMeshPrototype
import org.saar.core.mesh.prototype.VertexMeshPrototype
import org.saar.lwjgl.opengl.constants.DataType
import org.saar.lwjgl.opengl.constants.RenderMode
import org.saar.lwjgl.opengl.drawcall.ArraysDrawCall
import org.saar.lwjgl.opengl.drawcall.ElementsDrawCall
import org.saar.lwjgl.opengl.drawcall.InstancedArraysDrawCall
import org.saar.lwjgl.opengl.drawcall.InstancedElementsDrawCall
import org.saar.lwjgl.opengl.vao.Vao

object Meshes {

    @JvmStatic
    fun toArraysMesh(prototype: VertexMeshPrototype<*>, vertices: Int): DrawCallMesh {
        val vao = Vao.create()

        prototype.vertexBuffers.forEach { it.store(0) }
        prototype.vertexBuffers.forEach { it.loadInVao(vao) }

        return DrawCallMesh(vao, ArraysDrawCall(
            RenderMode.TRIANGLES, 0, vertices))
    }

    @JvmStatic
    fun toElementsMesh(prototype: IndexedVertexMeshPrototype<*>, indices: Int): Mesh {
        val vao = Vao.create()

        prototype.vertexBuffers.forEach { it.store(0) }
        prototype.vertexBuffers.forEach { it.loadInVao(vao) }

        prototype.indexBuffers.forEach { it.store(0) }
        prototype.indexBuffers.forEach { it.loadInVao(vao) }

        return DrawCallMesh(vao, ElementsDrawCall(
            RenderMode.TRIANGLES, indices, DataType.U_INT))
    }

    @JvmStatic
    fun toInstancedArrayMesh(
        prototype: InstancedMeshPrototype<*>, vertices: Int, instances: Int,
        renderMode: RenderMode = RenderMode.TRIANGLES,
    ): Mesh {
        val vao = Vao.create()

        prototype.instanceBuffers.forEach { it.store(0) }
        prototype.instanceBuffers.forEach { it.loadInVao(vao) }

        return DrawCallMesh(vao, InstancedArraysDrawCall(renderMode, vertices, instances))
    }

    @JvmStatic
    fun toInstancedElementsMesh(
        prototype: InstancedIndexedVertexMeshPrototype<*, *>,
        indices: Int,
        instances: Int,
    ): Mesh {
        val vao = Vao.create()

        prototype.vertexBuffers.forEach { it.store(0) }
        prototype.vertexBuffers.forEach { it.loadInVao(vao) }

        prototype.instanceBuffers.forEach { it.store(0) }
        prototype.instanceBuffers.forEach { it.loadInVao(vao) }

        prototype.indexBuffers.forEach { it.store(0) }
        prototype.indexBuffers.forEach { it.loadInVao(vao) }

        val drawCall = InstancedElementsDrawCall(
            RenderMode.TRIANGLES, indices, DataType.U_INT, instances)
        return DrawCallMesh(vao, drawCall)
    }
}